tests/test_qnode.py

```diff
@@ -14,377 +14,553 @@
 """
 Unit tests for the :mod:`pennylane` :class:`QNode` class.
 """
-import pytest
-import unittest
-import logging as log
-log.getLogger('defaults')
+import math
+from unittest.mock import Mock, PropertyMock, patch
 
-import autograd
+import pytest
 from autograd import numpy as np
 
-from defaults import pennylane as qml, BaseTest
+import pennylane as qml
+from pennylane._device import Device
+from pennylane.qnode import QNode, QuantumFunctionError, _flatten, unflatten
+
+flat_dummy_array = np.linspace(-1, 1, 64)
+
+
+class TestHelperMethods:
+    """Tests the internal helper methods of QNode"""
+
+    @pytest.mark.parametrize(
+        "shape",
+        [
+            (64,),
+            (64, 1),
+            (32, 2),
+            (16, 4),
+            (8, 8),
+            (16, 2, 2),
+            (8, 2, 2, 2),
+            (4, 2, 2, 2, 2),
+            (2, 2, 2, 2, 2, 2),
+        ],
```

Comment:
Since this test data is being re-used, may as well make a global variable at the top of the test file:
```python
TEST_ARRAY = [
        (64,),
        (64, 1),
        (32, 2),
        (16, 4),
        (8, 8),
        (16, 2, 2),
        (8, 2, 2, 2),
        (4, 2, 2, 2, 2),
        (2, 2, 2, 2, 2, 2),
]
```